lib/miou_unix.ml

```diff
@@ -367,25 +366,25 @@ let select uid (interrupt, p) ~block cancelled_syscalls =
       if interrupted interrupt rds then begin
         Logs.debug (fun m ->
             m "[%a] interrupted by Miou" Miou.Domain.Uid.pp uid);
-        intr interrupt p
+        intr interrupt
       end;
       let signals = collect domain [] in
       let signals = transmit_fds signals domain.revert domain.readers rds in
       let signals = transmit_fds signals domain.revert domain.writers wrs in
       signals
 
-let rec interrupt oc interrupted () =
-  if Atomic.get interrupted = 0 then
-    if Unix.single_write oc signal 0 1 = 0 then interrupt oc interrupted ()
-    else ignore (Atomic.fetch_and_add interrupted 1)
+let interrupt oc () =
+  match Unix.single_write oc signal 0 1 with
+  | _ -> ()
+  | exception Unix.(Unix_error (EAGAIN, _, _)) -> ()
 
 and signal = Bytes.make 1 '\000'
 
 let events domain =
   let ic, oc = Unix.pipe ~cloexec:true () in
```

Comment:
You can also make ic non-blocking, and handle EAGAIN on read as well, on the odd chance that something is slightly desynced, you prevent an erroneous block on read.